Fix background updates failing to add unique indexes on receipts

squahtx opened the pull request:
As part of the database migration to support threaded receipts, there is
a possible window in between
`73/08thread_receipts_non_null.sql.postgres` removing the original
unique constraints on `receipts_linearized` and `receipts_graph` and the
`reeipts_linearized_unique_index` and `receipts_graph_unique_index`
background updates from `72/08thread_receipts.sql` completing where
the unique constraints on `receipts_linearized` and `receipts_graph` are
missing. Any emulated upserts on these tables must therefore be
performed with a lock held, otherwise duplicate rows can end up in the
tables when there are concurrent emulated upserts. Fix the missing lock.

Note that emulated upserts no longer happen by default on sqlite, since
the minimum supported version of sqlite supports native upserts by
default now.

Finally, clean up any duplicate receipts that may have crept in before
trying to create the `receipts_graph_unique_index` and
`receipts_linearized_unique_index` unique indexes.

Signed-off-by: Sean Quah <seanq@matrix.org>

---

Resolves #14406, but not the follow on to remove `lock=False`.

The first commit fixes the race that creates the duplicates.
The second commit changes the background update to clean up the duplicates.

squahtx commented on synapse/storage/databases/main/receipts.py:
```diff
@@ -938,6 +924,86 @@ def _populate_receipt_event_stream_ordering_txn(
+            txn.execute(sql)
+            duplicate_keys = cast(List[Tuple[str, str, str]], list(txn))
+
+            # Then remove all duplicate receipts.
+            # We could be clever and try to keep the latest receipt out of every set of
+            # duplicates, but it's far simpler to remove them all.
+            for room_id, receipt_type, user_id in duplicate_keys:
+                sql = f"""
+                    DELETE FROM {table}
+                    WHERE
+                        room_id = ? AND
+                        receipt_type = ? AND
+                        user_id = ? AND
+                        thread_id IS NULL
+                """
+                txn.execute(sql, (room_id, receipt_type, user_id))
```
Is it worth trying to preserve these read receipts?

For `receipts_linearized`, we can choose to keep the receipt with the highest `stream_id`.
For `receipts_graph`, we have to dig into `data`, which is a user-provided json field, and extract out the `ts` field. This is a lot more painful, as we have to handle invalid data and ties.

clokep replied:
Although deleting from `receipts_linearized` *will* have a user-visible impact, so I'm unsure if that's a good idea.

squahtx replied:
I did a bit of rejigging to preserve the receipt with the highest `stream_id` in `receipts_linearized`. Users shouldn't notice anything now.

clokep commented on synapse/storage/databases/main/receipts.py:
```diff
@@ -113,24 +113,6 @@ def __init__(
             prefilled_cache=receipts_stream_prefill,
         )
 
-        self.db_pool.updates.register_background_index_update(
```
Why move these to the non-worker store?

squahtx replied:
I moved them to the `*BackgroundUpdateStore`. I thought that's where we usually put the background updates?
Is there a motivation for having these on the worker store that I've completely missed?

clokep replied:
There's no motivation besides it being consistent with other examples I saw. 🤷

clokep commented:
(I think someone non-me should review this since I made the original issue causing this. Overall I think the fix looks reasonable though.)

erikjohnston reviewed (approved):
Seems sane to me

DMRobertson commented on synapse/storage/databases/main/receipts.py:
```diff
@@ -702,9 +702,6 @@ def _insert_linearized_receipt_txn(
                 "data": json_encoder.encode(data),
             },
             where_clause=where_clause,
-            # receipts_linearized has a unique constraint on
-            # (user_id, room_id, receipt_type), so no need to lock
-            lock=False,
```
To double check: is the table correctly deemed safe to upsert into when the relevant background updates have run? (Wasn't sure how the second commit would affect this, if at all)

squahtx replied:
Yes, once the unique index has been added by the background update, we will be able to rely on native upserts again (and the value of `lock` won't matter).

squahtx commented:
This is failing trial on postgres with `psycopg2.errors.ActiveSqlTransaction: CREATE INDEX CONCURRENTLY cannot run inside a transaction block`

squahtx commented:
I've updated and refactored things a bit to
* fix the background update failing on postgres, because autocommit needs to be turned on when creating concurrent indexes.
* and keep the receipt with the highest `stream_id` in `receipt_linearized`
